display_list/display_list_builder.cc

```diff
@@ -135,9 +135,52 @@ void DisplayListBuilder::onSetImageFilter(sk_sp<SkImageFilter> filter) {
       : Push<ClearImageFilterOp>(0, 0);
 }
 void DisplayListBuilder::onSetColorFilter(sk_sp<SkColorFilter> filter) {
-  (current_color_filter_ = filter)  //
-      ? Push<SetColorFilterOp>(0, 0, std::move(filter))
-      : Push<ClearColorFilterOp>(0, 0);
+  current_color_filter_ = filter;
+  if (!filter) {
+    current_color_filter_type_ = kNone;
+    Push<ClearColorFilterOp>(0, 0);
+  } else {
```

Comment:
Would it be better to factor all of this logic out into our own ColorFilter class? We could have a ctor that takes an SkColorFilter and returns a ColorFilter that has this logic in it, and other ctors like ColorFilter(color, blendmode), ColorFilter(matrix)... etc. We would then store that object on the DisplayListBuilder and reference that in UpdateCurrentOpacityCompatibility().

I think that adding excessive logic like this to DisplayListBuilder makes the class more and more difficult to keep track of what's going on, especially if we are planning to start adding logic similar to this for other objects like ImageFilters and Shaders. Conceptually I see DisplayListBuilder as just a simple recorder/dispatcher class, but maybe that's a faulty view on my part.

Reply:
I have a local branch that I haven't turned into a PR yet where I do this. I create "display_list_color_filter.h/cc". There is a subclass that holds an SkColorFilter, but I don't have it do that logic in its constructor (because it's a constructor). I may make a factory static method, though, that encapsulates this logic.